lib/attr_extras/attr_initialize/attributes_declaration.rb

```diff
@@ -0,0 +1,44 @@
+module AttrExtras
+  InvalidParameterDeclaration = Class.new(StandardError)
+
+  class AttrInitialize
+    class AttributesDeclaration
+      attr_reader :names_with_array_of_hash_arguments
+      private :names_with_array_of_hash_arguments
+
+      def initialize(names_with_array_of_hash_arguments)
+        @names_with_array_of_hash_arguments = names_with_array_of_hash_arguments
+      end
+
+      def positional_attributes
+        return @attributes_names if @attributes_names
+
+        @attributes_names = names_with_array_of_hash_arguments.take_while { |name_or_array| !name_or_array.is_a?(Array) }
+        raise InvalidParameterDeclaration if (names_with_array_of_hash_arguments - @attributes_names).count > 1
+
+        @attributes_names
+      end
+
+      def hash_attributes
+        @hash_names ||= hash_names_without_default_values + default_values.keys
+      end
+
+      def default_values
+        return @default_values if @default_values
+
+        @default_values = (hash_names_with_default_values - hash_names_without_default_values).first
+        @default_values ||= {}
+      end
+
+      private
+
+      def hash_names_with_default_values
+        (names_with_array_of_hash_arguments - positional_attributes).flatten
+      end
+
+      def hash_names_without_default_values
+        hash_names_with_default_values.take_while { |name| !name.is_a?(Hash) }
```

Comment:
So this assumes all the hash names without defaults come first. The raise on line 17 only verifies that positional vs. named args come in the right order and so on, right? I suspect it won't catch something like `attr_initialize :foo, [ { bar: "baz" }, :bleh ]`? But maybe that's too unlikely?